pkg/controllers/management/clustertemplate/revision_controller.go

```diff
@@ -0,0 +1,118 @@
+package clustertemplate
+
+import (
+	"context"
+
+	"fmt"
+
+	"strings"
+
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	ClusterTemplateRevController = "mgmt-cluster-template-rev-controller"
+)
+
+type CTemplateRevController struct {
+	clusterTemplates              v3.ClusterTemplateInterface
+	clusterTemplateLister         v3.ClusterTemplateLister
+	clusterTemplateRevisions      v3.ClusterTemplateRevisionInterface
+	clusterTemplateRevisionLister v3.ClusterTemplateRevisionLister
+	managementContext             *config.ManagementContext
+}
+
+func newCTemplateRevController(ctx context.Context, mgmt *config.ManagementContext) *CTemplateRevController {
+	n := &CTemplateRevController{
+		clusterTemplates:              mgmt.Management.ClusterTemplates(namespace.GlobalNamespace),
+		clusterTemplateLister:         mgmt.Management.ClusterTemplates(namespace.GlobalNamespace).Controller().Lister(),
+		managementContext:             mgmt,
+		clusterTemplateRevisions:      mgmt.Management.ClusterTemplateRevisions(namespace.GlobalNamespace),
+		clusterTemplateRevisionLister: mgmt.Management.ClusterTemplateRevisions(namespace.GlobalNamespace).Controller().Lister(),
+	}
+	return n
+}
+
+func Register(ctx context.Context, management *config.ManagementContext) {
+	n := newCTemplateRevController(ctx, management)
+	if n != nil {
+		management.Management.ClusterTemplateRevisions("").AddHandler(ctx, ClusterTemplateRevController, n.sync)
+	}
+}
+
+//sync is called periodically and on real updates
+func (n *CTemplateRevController) sync(key string, obj *v3.ClusterTemplateRevision) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	//check if this revision has isDefault = true, if yes update the parent Template
```

Comment:
Do we ever validate that there is only one template revision with default attribute set to true? On the API level

Reply:
@alena1108 I feel we should not have the isDefault flag on templateRevision, reason:
1)Right, we need to validate on API level that only one revision has isDefault true. This check will need to be done for both Create and Update templateRevision APIs
2) Also, for  Update templateRevision API, we are going to error out if some cluster is using this templateRevision.
3) Now when Update templateRevision API is called, to satisfy #1 and #2 above, we will have to check if isDefault is getting updated and do check #1. But for #2, we will have to scan the data and find if anything else is getting updated - and this is very inefficient.
4) Moreover, we can never guard against concurrent API requests in HA mode.

So I think we should not have this flag on templateRevision. Just having ClusterTemplate.DefaultRevision reference should be enough. Does this sound ok?

Reply:
I agree with your your points @prachidamle; the resolution sounds reasonable. We'll only have to run it with the UI team. TemplateId will always be set on a cluster, so I guess it should be easy to fetch the default templateRevision if needed.